plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -422,11 +473,339 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+// many of them as possible while checking that the "pod" can still fit on the node.
+// NOTE: This function assumes that it is never called if "pod" cannot be scheduled
+// due to pod affinity, node affinity, or node anti-affinity reasons. None of
+// these predicates can be satisfied by removing more pods from the node.
+// TODO(bsalamat): Add support for PodDisruptionBudget.
+func selectVictimsOnNode(
+	pod *v1.Pod,
+	meta algorithm.PredicateMetadata,
+	nodeInfo *schedulercache.NodeInfo,
+	fitPredicates map[string]algorithm.FitPredicate,
+	extenders []algorithm.SchedulerExtender) ([]*v1.Pod, bool) {
+	higherPriority := func(pod1, pod2 interface{}) bool {
+		return util.GetPodPriority(pod1.(*v1.Pod)) > util.GetPodPriority(pod2.(*v1.Pod))
+	}
+	potentialVictims := util.SortableList{CompFunc: higherPriority}
+	nodeInfoCopy := nodeInfo.Clone()
+
+	removePod := func(rp *v1.Pod) {
+		nodeInfoCopy.RemovePod(rp)
+		if meta != nil {
+			meta.RemovePod(rp)
+		}
+	}
+	addPod := func(ap *v1.Pod) {
+		nodeInfoCopy.AddPod(ap)
+		if meta != nil {
+			meta.AddPod(ap, nodeInfoCopy)
+		}
+	}
+	// As the first step, remove all the lower priority pods from the node and
+	// check if the given pod can be scheduled.
+	podPriority := util.GetPodPriority(pod)
+	for _, p := range nodeInfoCopy.Pods() {
+		if util.GetPodPriority(p) < podPriority {
+			potentialVictims.Items = append(potentialVictims.Items, p)
+			removePod(p)
+		}
+	}
+	potentialVictims.Sort()
+	// If the new pod does not fit after removing all the lower priority pods,
+	// we are almost done and this node is not suitable for preemption. The only condition
+	// that we should check is if the "pod" is failing to schedule due to pod affinity
+	// failure.
+	if fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+		if err != nil {
+			glog.Warningf("Encountered error while selecting victims on node %v: %v", nodeInfo.Node().Name, err)
+			return nil, false
+		}
+		// If the new pod still cannot be scheduled for any reason other than pod
+		// affinity, the new pod will not fit on this node and we are done here.
+		affinity := pod.Spec.Affinity
+		if affinity == nil || affinity.PodAffinity == nil {
+			return nil, false
+		}
+		for _, failedPred := range failedPredicates {
+			if failedPred != predicates.ErrPodAffinityNotMatch {
+				return nil, false
+			}
+		}
+		// If we reach here, it means that the pod cannot be scheduled due to pod
+		// affinity or anti-affinity. Since failure reason for both affinity and
+		// anti-affinity is the same, we cannot say which one caused it. So, we try
+		// adding pods one at a time and see if any of them satisfies the affinity rules.
+		for i, p := range potentialVictims.Items {
+			existingPod := p.(*v1.Pod)
+			addPod(existingPod)
+			if fits, _, _ = podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+				removePod(existingPod)
```

Comment:
That is not correct.
A pod may potentially require more than one of the pod that you already removed to be schedulable.

Actually, computing set of those may be pretty tricky.

Reply:
> A pod may potentially require more than one of the pod that you already removed to be schedulable.

This is what we thought at first but I don't think it's true (unless we're missing something).
My recollection (I need to go back and check the code) is that pod affinity works like "Pod X can only run on a node (/rack/zone but larger-than-node is not relevant) that has at least one pod matching the following label selector expression." So when you are scheduling a new pod, either it is infeasible even with all of the existing pods, or it requires none of the existing pods, or it requires some one of the existing pods, but it never requires more than one of them. But we should double-check.

Reply:
Yeah - that makes sense to me.
@bsalamat - can you please double check that and if that's really true (I think it is), please add a comment explaining it? I think we will forget about it in few months

Reply:
I thought about it a bit more and I don't agree.

What you wrote is true if there is a single affinity term. But there can be multiple of those and they can be targeting different pods. So I still actually believe that this is not correct wrt pod affinity.

Reply:
@bsalamat please check the affinity implementation. @wojtek-t may be right.

Reply:
Looks like @wojtek-t is right. [This loop](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L1174) checks each pod affinity term individually, and admits the pod if each term can individually be matched by some other pod. The comment in the API is confusing and I forgot how this works. Oops.

Reply:
You are right. That makes finding the minimum set of pods very computationally very expensive, as we should try all permutations of victims.

Reply:
Actually, I think I have an idea that can solve the problem without checking permutations. Let me try.

Reply:
I think I addressed the problem and added an integration test to verify it.

Reply:
I have been thinking more about this new algorithm. It solves the problem of having multiple affinity rules, each satisfied by a different pod, but it doesn't solve the case where a pod has both affinity and anti-affinity rules and some of the existing pods must be preempted to satisfy the anti-affinity rules and some other existing pods must be saved to satisfy the affinity rules.

Reply:
Thinking even more about this, it makes little sense for a pod to have affinity on a lower priority pod, because that lower priority pod may get preempted any moment. So, pods should have affinity on same or higher priority pods. If we set this restriction that preemption does not support affinity on lower priority pods, the logic will become much simpler. I think we should consider it at least for the first version.